File: docs/snippets/ov_properties_migration.py
# Copyright (C) 2018-2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
#

import openvino as ov
import openvino.properties as props
import openvino.properties.hint as hints
import openvino.properties.device as device
import openvino.properties.streams as streams

from utils import get_model

def main():
    core = ov.Core()

    # ! [core_set_property]
    core.set_property(device_name="CPU", properties={props.enable_profiling: True})
    # ! [core_set_property]

    model = get_model()

    if "GPU" not in core.available_devices:
        return 0

    # ! [core_compile_model]
    compiled_model = core.compile_model(model=model, device_name="MULTI", config=
        {
            device.priorities: "GPU,CPU",
            hints.performance_mode: hints.PerformanceMode.THROUGHPUT,
            hints.inference_precision: ov.Type.f32
        })
    # ! [core_compile_model]

    # ! [compiled_model_set_property]
    # turn CPU off for multi-device execution
    compiled_model.set_property(properties={device.priorities: "GPU"})
    # ! [compiled_model_set_property]

    # ! [core_get_rw_property]
    num_streams = core.get_property("CPU", streams.num)
    # ! [core_get_rw_property]

    # ! [core_get_ro_property]
    full_device_name = core.get_property("CPU", device.full_name)
    # ! [core_get_ro_property]

    # ! [compiled_model_get_rw_property]
    perf_mode = compiled_model.get_property(hints.performance_mode)
    # ! [compiled_model_get_rw_property]

    # ! [compiled_model_get_ro_property]
    nireq = compiled_model.get_property(props.optimal_number_of_infer_requests)
    # ! [compiled_model_get_ro_property]

    import ngraph as ng
    import openvino.inference_engine as ie
    from utils import get_ngraph_model

    core = ie.IECore()
    #! [core_get_metric]
    full_device_name = core.get_metric("CPU", "FULL_DEVICE_NAME")
    #! [core_get_metric]

    #! [core_get_config]
    num_streams = core.get_config("CPU", "CPU_THROUGHPUT_STREAMS")
    #! [core_get_config]

    #! [core_set_config]
    core.set_config({"PERF_COUNT": "YES"}, "CPU")
    #! [core_set_config]

    net = get_ngraph_model()

    #! [core_load_network]
    exec_network = core.load_network(net, "MULTI", {"DEVICE_PRIORITIES": "CPU, GPU",
                                                    "PERFORMANCE_HINT": "THROUGHPUT",
                                                    "ENFORCE_BF16": "NO"})
    #! [core_load_network]

    #! [executable_network_set_config]
    # turn CPU off for multi-device execution
    exec_network.set_config({"DEVICE_PRIORITIES": "GPU"})
    #! [executable_network_set_config]

    #! [executable_network_get_metric]
    nireq = exec_network.get_metric("OPTIMAL_NUMBER_OF_INFER_REQUESTS")
    #! [executable_network_get_metric]

    #! [executable_network_get_config]
    perf_hint = exec_network.get_config("PERFORMANCE_HINT")
    #! [executable_network_get_config]

File: docs/snippets/ov_properties_migration.cpp
#include <openvino/runtime/core.hpp>
#ifndef IN_OV_COMPONENT
#    define IN_OV_COMPONENT
#    define WAS_OV_LIBRARY_DEFINED
#endif

#include <inference_engine.hpp>

#ifdef WAS_OV_LIBRARY_DEFINED
#    undef IN_OV_COMPONENT
#    undef WAS_OV_LIBRARY_DEFINED
#endif

int main_new() {
    ov::Core core;

//! [core_get_ro_property]
// 'auto' is automatically deduced as std::string
// since the type is stored in the property
auto full_device_name = core.get_property("CPU", ov::device::full_name);
//! [core_get_ro_property]

//! [core_get_rw_property]
// 'auto' is automatically deduced as ov::streams::Num
// since the type is stored in the property
auto num_streams = core.get_property("CPU", ov::streams::num);
//! [core_get_rw_property]

//! [core_set_property]
core.set_property("CPU", ov::enable_profiling(true));
//! [core_set_property]

auto model = core.read_model("sample.xml");
//! [core_compile_model]
auto compiled_model = core.compile_model(model, "MULTI",
    ov::device::priorities("GPU", "CPU"),
    ov::hint::performance_mode(ov::hint::PerformanceMode::THROUGHPUT),
    ov::hint::inference_precision(ov::element::f32));
//! [core_compile_model]

//! [compiled_model_set_property]
// turn CPU off for multi-device execution
compiled_model.set_property(ov::device::priorities("GPU"));
//! [compiled_model_set_property]

{
//! [compiled_model_get_ro_property]
// 'auto' is deduced to 'uint32_t'
auto nireq = compiled_model.get_property(ov::optimal_number_of_infer_requests);
//! [compiled_model_get_ro_property]
}

{
//! [compiled_model_get_rw_property]
ov::hint::PerformanceMode perf_mode = compiled_model.get_property(ov::hint::performance_mode);
//! [compiled_model_get_rw_property]
}


return 0;
}


int main_old() {
    InferenceEngine::Core core;
//! [core_get_metric]
auto full_device_name = core.GetMetric("CPU", METRIC_KEY(FULL_DEVICE_NAME)).as<std::string>();
//! [core_get_metric]

//! [core_get_config]
// a user has to parse std::string after
auto num_streams = core.GetConfig("CPU", CONFIG_KEY(CPU_THROUGHPUT_STREAMS)).as<std::string>();
//! [core_get_config]

//! [core_set_config]
core.SetConfig({ { CONFIG_KEY(PERF_COUNT), CONFIG_VALUE(YES) } }, "CPU");
//! [core_set_config]

auto model = core.ReadNetwork("sample.xml");
//! [core_load_network]
auto exec_network = core.LoadNetwork(model, "MULTI", {
    { MULTI_CONFIG_KEY(DEVICE_PRIORITIES), "CPU, GPU" },
    { CONFIG_KEY(PERFORMANCE_HINT), CONFIG_VALUE(THROUGHPUT) },
    { CONFIG_KEY(ENFORCE_BF16), CONFIG_VALUE(NO) } });
//! [core_load_network]

//! [executable_network_set_config]
// turn CPU off for multi-device execution
exec_network.SetConfig({ { MULTI_CONFIG_KEY(DEVICE_PRIORITIES), "GPU" } });
//! [executable_network_set_config]

{
//! [executable_network_get_metric]
auto nireq = exec_network.GetMetric(EXEC_NETWORK_METRIC_KEY(OPTIMAL_NUMBER_OF_INFER_REQUESTS)).as<uint32_t>();
//! [executable_network_get_metric]
}

{
//! [executable_network_get_config]
std::string perf_model = exec_network.GetConfig(CONFIG_KEY(PERFORMANCE_HINT)).as<std::string>();
//! [executable_network_get_config]
}

return 0;
}
